types/react/index.d.ts

```diff
@@ -925,7 +925,7 @@ declare namespace React {
      * @version 16.8.0
      * @see https://reactjs.org/docs/hooks-reference.html#usestate
      */
-    function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];
+    function useState<S>(initialState: (S extends ((...args: any[]) => any) ? never : S) | (() => S)): [S, Dispatch<SetStateAction<S>>];
```

Comment:
```
function useState<S extends (...args: any[]) => any)>(initialState: () => S): [S, Dispatch<(prevState: S) => S>];

// then keep
function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];
```

Will this work with [test case](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57563/commits/3fd05247a2bb3414a5be51ebda8d923ea8a0bef7#diff-ee56bea8f592b38602c6e3e0339fec47825c5397221f103cecd5e37bbdc6ecc7R39-R42) passed?

Reply:
<img width="735" alt="image" src="https://user-images.githubusercontent.com/12760587/154208770-7a3e6336-0d24-4877-83f5-4d5d40c65bb2.png">

previous test case passes if keep this line, but then this PR's main purpose can't be done

```tsx
const initialValue: (() => number) = something;
const [a, setA] = useState<() => number>(initialValue); // no type error here :'(
```